File: 世界で戦うプログラミング力を鍛える150問/Chapter_1/Chapter_1_etc.php
<?php

class Student
{
    public int $id;
    public string $name;

    public function __construct(int $id, string $name)
    {
        $this->id = $id;
        $this->name = $name;
    }
}

function buildMap(array $students): array
{
    $map = [];

    foreach ($students as $student) {
        $map[$student->id] = $student;
    }

    return $map;
}

$students = [
    new Student(1, "山田"),
    new Student(2, "鈴木"),
    new Student(3, "佐藤")
];

// 配列検索 O(n)
$fromArray = null;

foreach ($students as $student) {
    if ($student->id === 3) {
        $fromArray = $student;
        break;
    }
}

echo "配列検索\n";
print_r($fromArray);

// HashMap検索 O(1)
$map = buildMap($students);

echo "\nHashMap検索\n";
print_r($map[3]);

File: 世界で戦うプログラミング力を鍛える150問/Chapter_1/Chapter_1_etc.kt
data class Student(
    val id: Int,
    val name: String
)

fun buildMap(students: Array<Student>): HashMap<Int, Student> {

    val map = HashMap<Int, Student>()

    for (student in students) {
        map[student.id] = student
    }

    return map
}

fun main() {

    val students = arrayOf(
        Student(1, "山田"),
        Student(2, "鈴木"),
        Student(3, "佐藤")
    )

    // 配列検索 O(n)
    val fromArray = students.firstOrNull {
        it.id == 3
    }

    println("配列検索")
    println(fromArray)

    // HashMap検索 O(1)
    val map = buildMap(students)

    println("\nHashMap検索")
    println(map[3])
}